classes/form/CustomerAddressForm.php

```diff
@@ -217,4 +203,111 @@ function (FormField $item) {
             'formFields' => $formFields,
         ];
     }
+
+    /**
+     * Performs validation on field values.
+     * Returns true if all field values are correct, false otherwise.
+     *
+     * @return bool
+     */
+    private function validateFieldsValues(): bool
+    {
+        $isValid = true;
+
+        $isValid &= $this->validatePostcode();
+        $isValid &= $this->validateFirstname();
+        $isValid &= $this->validateLastname();
+        $isValid &= $this->validateCity();
+
+        return (bool) $isValid;
+    }
+
+    private function validatePostcode(): bool
+    {
+        if (($postcode = $this->getField('postcode'))) {
+            if ($postcode->isRequired()) {
+                $country = $this->formatter->getCountry();
+                if (!$country->checkZipCode($postcode->getValue())) {
+                    $postcode->addError($this->translator->trans(
+                        'Invalid postcode - should look like "%zipcode%"',
+                        ['%zipcode%' => $country->zip_code_format],
+                        'Shop.Forms.Errors'
+                    ));
+
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    private function validateCity(): bool
```

Comment:
Seems very close to `CustomerForm::validateLastname()`
- fieldname change
- there is some checks on `$field->isRequired()`
- `isCityName()` is used instead of `isCustomerName()`

So we could use a unique function for both usecases?
```
public function validateField(string $fieldName, $mustBeRequired = true, $validationFunction)

...

$isValid &= $this->validateField('city', true, 'isCityName');
```

Reply:
I thought exactly the same but making it over generic will make it hard to understand and if we want to add a specific case for one of the fields, we will have to break the generic
But why not? I can apply your suggestion ;)

Reply:
@sowbiba If you are thinking "ok today I can do X, but maybe in future X will not be good if some changes require to be done"... then I answer [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) 😉 

> "You aren't gonna need it" is a principle of extreme programming (XP) that states a programmer should not add functionality until deemed necessary. XP co-founder Ron Jeffries has written: "Always implement things when you actually need them, never when you just foresee that you need them." Other forms of the phrase include "You aren't going to need it" and "You ain't gonna need it".

Reply:
👍🏾 And that is a principle of agility : do what you need today
There is another difference between 'city' and the names : The error message
I can give it as parameter too :D